client/fs_endpoint.go

```diff
@@ -0,0 +1,921 @@
+		// interested in so we can stop there.
+		maxIndex := int64(math.MaxInt64)
+		if !follow {
+			_, idx, _, err := findClosest(entries, maxIndex, 0, task, logType)
+			if err != nil {
+				return err
+			}
+			maxIndex = idx
+		}
+
+		logEntry, idx, openOffset, err := findClosest(entries, nextIdx, offset, task, logType)
+		if err != nil {
+			return err
+		}
+
+		var eofCancelCh chan error
+		exitAfter := false
+		if !follow && idx > maxIndex {
+			// Exceeded what was there initially so return
+			return nil
+		} else if !follow && idx == maxIndex {
+			// At the end
+			eofCancelCh = make(chan error)
+			close(eofCancelCh)
+			exitAfter = true
+		} else {
+			eofCancelCh = blockUntilNextLog(ctx, fs, logPath, task, logType, idx+1)
+		}
+
+		p := filepath.Join(logPath, logEntry.Name)
+		err = f.streamFile(ctx, openOffset, p, 0, fs, framer, eofCancelCh)
+
+		// Check if the context is cancelled
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+
+		if err != nil {
+			// Check if there was an error where the file does not exist. That means
+			// it got rotated out from under us.
+			if os.IsNotExist(err) {
+				continue
+			}
+
+			// Check if the connection was closed
+			if err == syscall.EPIPE {
+				return nil
+			}
+
+			return fmt.Errorf("failed to stream %q: %v", p, err)
+		}
+
+		if exitAfter {
+			return nil
+		}
+
+		// defensively check to make sure StreamFramer hasn't stopped
+		// running to avoid tight loops with goroutine leaks as in
+		// #3342
+		select {
+		case <-framer.ExitCh():
+			err := parseFramerErr(framer.Err())
+			if err == syscall.EPIPE {
+				// EPIPE just means the connection was closed
+				return nil
+			}
+			return err
+		default:
+		}
+
+		// Since we successfully streamed, update the overall offset/idx.
+		offset = int64(0)
+		nextIdx = idx + 1
+	}
+}
+
+// streamFile is the internal method to stream the content of a file. If limit
+// is greater than zero, the stream will end once that many bytes have been
+// read. eofCancelCh is used to cancel the stream if triggered while at EOF. If
+// the connection is broken an EPIPE error is returned
+func (f *FileSystem) streamFile(ctx context.Context, offset int64, path string, limit int64,
+	fs allocdir.AllocDirFS, framer *sframer.StreamFramer, eofCancelCh chan error) error {
+
+	// Get the reader
+	file, err := fs.ReadAt(path, offset)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	var fileReader io.Reader
+	if limit <= 0 {
+		fileReader = file
+	} else {
+		fileReader = io.LimitReader(file, limit)
+	}
+
+	// Create a tomb to cancel watch events
+	waitCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Create a variable to allow setting the last event
+	var lastEvent string
+
+	// Only create the file change watcher once. But we need to do it after we
+	// read and reach EOF.
+	var changes *watch.FileChanges
+
+	// Start streaming the data
+	data := make([]byte, streamFrameSize)
```

Comment:
can we set this to min(limit, streamFrameSize) to prevent allocating an array larger than we need when limit < streamFrameSize?